polar-core/src/debugger.rs

```diff
@@ -172,23 +111,51 @@ impl Debugger {
     /// - `Some(Goal::Debug { message })` -> Pause evaluation.
     /// - `None` -> Continue evaluation.
     fn maybe_break(&self, event: DebugEvent, vm: &PolarVirtualMachine) -> Option<Rc<Goal>> {
-        self.step.as_ref().and_then(|step| match (step, event) {
-            (Step::Goal, DebugEvent::Goal(goal)) => Some(Rc::new(Goal::Debug {
-                message: goal.to_string(),
-            })),
-            (Step::Over { snapshot }, DebugEvent::Query)
-            | (Step::Out { snapshot }, DebugEvent::Query)
-                if vm.queries[..vm.queries.len() - 1] == snapshot[..] =>
+        if let Some(step) = self.step.as_ref() {
+            match (step, event) {
+                (Step::Goal, DebugEvent::Goal(goal)) => Some(Rc::new(Goal::Debug {
+                    message: goal.to_string(),
+                })),
+                (Step::Into, DebugEvent::Query) => self.break_query(vm),
+                (Step::Out { level }, DebugEvent::Query)
+                    if vm.trace_stack.is_empty() || vm.trace_stack.len() < *level =>
+                {
+                    self.break_query(vm)
+                }
+                (Step::Over { level }, DebugEvent::Query) if vm.trace_stack.len() == *level => {
+                    self.break_query(vm)
+                }
+                _ => None,
+            }
+        } else {
+            None
+        }
```

Comment:
nit: The previous format (an `.and_then()` instead of an `if let Some(step)`) obviates this trailing `else` block